File: Challenges/removeOuterParanthesis.py
class Solution:
    def removeOuterParentheses(self, S: str) -> str:
        stack = []
        res = []
        count = 0
        for p in S:
            if p == '(':
                stack.append(p) # increment count of open parenthesis
                count += 1
            else:
                count -= 1 # decrement count of open parenthesis
                # we only append when there is more than one opening parenthesis
                # because we want to remove the outer most
                if len(stack) > 1:
                    stack.append(p)
                else:
                    stack.pop()
            if not count:
                if len(stack) > 2: # remove outer paranthesis
                    stack.pop(0)
                    stack.pop()
                res.extend(stack)
                stack = []
        if len(stack) > 2:
            stack.pop(0)
            stack.pop()
        res.extend(stack)
        return ''.join(res)

File: Challenges/removeOuterParanthesis.java
class Solution {
    public String removeOuterParentheses(String S) {
        List<Integer> l = new ArrayList<>();
        Stack<Character> stack = new Stack<>();
        char[] arr = S.toCharArray();
        for (int i=0; i < arr.length; i++) {
            if (stack.size() == 0) {
                l.add(i);
                stack.push(arr[i]);
                continue;
            }
            if (stack.peek() == '(' && arr[i] == ')') {
                stack.pop();
                if (stack.size() == 0) {
                    l.add(i);
                } 
            } else if (stack.peek() == ')' && arr[i] == '(') {
                stack.pop();
                if (stack.size() == 0) {
                    l.add(i);
                }
            } else {
                stack.push(arr[i]);
            }
        }
        
        StringBuilder sb = new StringBuilder();
        for (int i=0; i<l.size(); i+=2) {
            sb.append(S.substring(l.get(i)+1, l.get(i+1)));
        }
        return sb.toString();
    }
}
